janusgraph-cql/src/test/java/org/janusgraph/graphdb/cql/CQLGraphTest.java

```diff
@@ -252,6 +359,42 @@ public void testLimitBatchSizeForMultiQueryMultiCardinalityProperties(boolean sl
         assertEquals(1, countBackendQueriesOfSize(lastBatch + lastBatch * 4 + lastBatch * 4, profile.getMetrics()));
     }
 
+    @MethodSource("generateGroupingConfigs")
+    @ParameterizedTest
+    public void testBatchWithCQLGrouping(int elementsAmount, Object[] configuration){
+
+        mgmt.makeVertexLabel("testVertex").make();
+        mgmt.makePropertyKey("singleProperty1").cardinality(Cardinality.SINGLE).dataType(String.class).make();
+        mgmt.makePropertyKey("singleProperty2").cardinality(Cardinality.SINGLE).dataType(String.class).make();
+        mgmt.makePropertyKey("singleProperty3").cardinality(Cardinality.SINGLE).dataType(String.class).make();
+        mgmt.makePropertyKey("setProperty").cardinality(Cardinality.SET).dataType(String.class).make();
+        mgmt.makePropertyKey("listProperty").cardinality(Cardinality.LIST).dataType(String.class).make();
+        finishSchema();
+        JanusGraphVertex[] cs = new JanusGraphVertex[elementsAmount];
+        for (int i = 0; i < elementsAmount; ++i) {
+            cs[i] = graph.addVertex("testVertex");
+            cs[i].property("singleProperty1", "single value1 "+i);
+            cs[i].property("singleProperty2", "single value1 "+i);
+            cs[i].property("singleProperty3", "single value1 "+i);
+            cs[i].property(VertexProperty.Cardinality.set, "setProperty", "setValue1");
+            cs[i].property(VertexProperty.Cardinality.set, "setProperty", "setValue2");
+            cs[i].property(VertexProperty.Cardinality.set, "setProperty", "setValue3");
+            cs[i].property(VertexProperty.Cardinality.set, "setProperty", "setValue4");
+            cs[i].property(VertexProperty.Cardinality.list, "listProperty", "listValue1");
+            cs[i].property(VertexProperty.Cardinality.list, "listProperty", "listValue2");
+            cs[i].property(VertexProperty.Cardinality.list, "listProperty", "listValue3");
+            cs[i].property(VertexProperty.Cardinality.list, "listProperty", "listValue4");
+        }
+
+        newTx();
+
+        clopen(configuration);
+
+        // test batching for `values()`
+        TraversalMetrics profile = testLimitedBatch(() -> graph.traversal().V(cs).barrier(elementsAmount).values("singleProperty1", "singleProperty2", "singleProperty3", "setProperty", "listProperty"), configuration);
```

Comment:
I have two questions:

1. `testLimitedBatch` only compares results with and without batching. In the context of this PR, we would want to compare the results with and without key/column grouping.
2. Can one have one query that requests different properties (columns) from different vertices (keys)? For example, `g.V('v1').values('prop1').as('v1Prop1').V('v2').values('prop2').as('v2Prop2').select('v1Prop1', 'v2Prop2')`. I am not sure if a query like this will trigger your new logic at all, but I would like to make sure the grouping doesn't assume we request the same columns for different keys.

Reply:
1. That makes sense. I will change to use a new method which compares results also with and without keys grouping.
2. This logic is applied to a single query like ‘values(“prop1”, “prop2”)’ only because that’s how multi-query is working. However, we may have some data already cached for some vertices in transaction level cache of db-level cache which will modify what we are actually querying from the storage backend. Nevertheless, those will be different query groups. The grouping logic is aware that there might be different columns asked for different set of keys (thats what query group means there). Grouping logic won’t ever fetch columns which were not explicitly requested for specific keys.

Reply:
> Nevertheless, those will be different query groups. The grouping logic is aware that there might be different columns asked for different set of keys (thats what query group means there)

Cool 👍

Reply:
Updated this test to test each configuration with enabled / disabled grouping.